chalk-engine/src/boring_impls.rs

```diff
@@ -0,0 +1,199 @@
+        Ok(FlounderedSubgoal {
+            floundered_literal: floundered_literal.fold_with(folder, outer_binder)?,
+            floundered_time: floundered_time.fold_with(folder, outer_binder)?,
+        })
+    }
+}
+
+impl<I: Interner, TI: TargetInterner<I>> Fold<I, TI> for Literal<I> {
+    type Result = Literal<TI>;
+
+    fn fold_with<'i>(
+        &self,
+        folder: &mut dyn Folder<'i, I, TI>,
+        outer_binder: DebruijnIndex,
+    ) -> Fallible<Self::Result>
+    where
+        I: 'i,
+        TI: 'i,
+    {
+        match self {
+            Literal::Positive(goal) => Ok(Literal::Positive(goal.fold_with(folder, outer_binder)?)),
+            Literal::Negative(goal) => Ok(Literal::Negative(goal.fold_with(folder, outer_binder)?)),
+        }
+    }
+}
+
+impl<I: Interner + Debug> Visit<I> for ExClause<I> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        let ExClause {
+            subst,
+            ambiguous: _,
+            constraints,
+            subgoals,
+            delayed_subgoals,
+            answer_time,
+            floundered_subgoals,
+        } = self;
+
+        R::new()
+            .and_then(|| subst.visit_with(visitor, outer_binder))
+            .and_then(|| constraints.visit_with(visitor, outer_binder))
+            .and_then(|| constraints.visit_with(visitor, outer_binder))
+            .and_then(|| subgoals.visit_with(visitor, outer_binder))
+            .and_then(|| delayed_subgoals.visit_with(visitor, outer_binder))
+            .and_then(|| answer_time.visit_with(visitor, outer_binder))
+            .and_then(|| floundered_subgoals.visit_with(visitor, outer_binder))
+    }
+}
+
+impl<I: Interner + Debug> Visit<I> for FlounderedSubgoal<I>
+where
+    InEnvironment<Constraint<I>>: Visit<I>,
+    Canonical<ConstrainedSubst<I>>: Visit<I>,
+{
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        let FlounderedSubgoal {
+            floundered_literal,
+            floundered_time,
+        } = self;
+
+        R::new()
+            .and_then(|| floundered_literal.visit_with(visitor, outer_binder))
+            .and_then(|| floundered_time.visit_with(visitor, outer_binder))
+    }
+}
+
+impl<I: Interner> Visit<I> for Literal<I> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        match self {
+            Literal::Positive(goal) => goal.visit_with(visitor, outer_binder),
+            Literal::Negative(goal) => goal.visit_with(visitor, outer_binder),
+        }
+    }
+}
+
+// FIXME: I can't figure out macro visibility rules
+macro_rules! copy_fold {
```

Comment:
I could use some help here. Just doing `chalk_ir::folder::copy_fold!(TableIndex)` gives the error `macro copy_fold is private`

Reply:
`#[macro_export]` is used to make macros public: https://doc.rust-lang.org/reference/macros-by-example.html#scoping-exporting-and-importing

Reply:
`copy_fold` does have `macro_export`: https://github.com/rust-lang/chalk/blob/master/chalk-ir/src/fold/boring_impls.rs#L220

Reply:
Try invoking `copy_fold` using `chalk_ir::copy_fold!` instead of `chalk_ir::folder::copy_fold`.

Basically using the `macro_export` attribute actually exports the macro at the root of the crate, rather than the module where it's defined. I'm not exactly *sure* why it works this way though, because it's rather unintuitive.

Reply:
Thanks